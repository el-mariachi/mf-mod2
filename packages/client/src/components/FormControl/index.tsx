import { FC } from 'react'
import Form from 'react-bootstrap/Form'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'
import { UseFormRegister, FieldErrorsImpl } from 'react-hook-form'

type FormControlProps = {
  formName: string
  controlProps: ControlProps
  register: UseFormRegister<Fields>
  errors: Partial<FieldErrorsImpl>
  readOnly?: boolean
}

type Fields = FormControlProps['controlProps']['name']

const FormControl: FC<FormControlProps> = ({
  formName,
  register,
  errors,
  controlProps,
  readOnly,
}) => {
  return (
    <Form.Group
      as={Row}
      className="mb-3"
      controlId={`${formName}-${controlProps.name}`}>
      <Form.Label column sm="3">
        {controlProps.label}
      </Form.Label>
      <Col sm={9}>
        <Form.Control
          type={controlProps.type}
          isInvalid={errors[controlProps.name] !== undefined}
          {...register(controlProps.name, {
            required: 'Поле должно быть заполнено',
            pattern: {
              value: controlProps.test,
              message: controlProps.message,
            },
          })}
          readOnly={readOnly}
        />
        <Form.Control.Feedback type="invalid">
          {String(errors[controlProps.name]?.message)}
        </Form.Control.Feedback>
      </Col>
    </Form.Group>
  )
}

export default FormControl
