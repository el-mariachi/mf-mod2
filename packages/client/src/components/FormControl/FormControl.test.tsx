/* eslint-disable @typescript-eslint/no-empty-function */
/* eslint-disable @typescript-eslint/no-explicit-any */
import FormControl from './index'
import { render } from '@testing-library/react'
import * as renderer from 'react-test-renderer'

describe('#FormControl', () => {
  const registerCallbackStub = () => {}
  it('FormControl should to match snapshot', () => {
    const snapshot = renderer
      .create(
        <FormControl
          formName={'formName'}
          controlProps={{
            name: 'name',
            label: 'label',
            type: 'type',
          }}
          register={registerCallbackStub as any}
          errors={[123] as any}
        />
      )
      .toJSON()
    expect(snapshot).toMatchSnapshot()
  })

  it('should create', () => {
    const renderedFormControl = render(
      <FormControl
        formName={'formName'}
        controlProps={{
          name: 'name',
          label: 'label',
          type: 'type',
        }}
        register={registerCallbackStub as any}
        errors={[123] as any}
      />
    )
    expect(renderedFormControl).toBeTruthy()
  })
  it('prop formName should be applied to label and input', () => {
    const { container } = render(
      <FormControl
        formName={'testName'}
        controlProps={{
          name: 'name',
          label: 'label',
          type: 'type',
        }}
        register={registerCallbackStub as any}
        errors={[123] as any}
      />
    )
    const label = container.querySelector('label')
    expect(label?.htmlFor).toBe('testName-name')
    const input = container.querySelector('input')
    expect(input?.id).toBe('testName-name')
  })
  it('prop name should be applied to label and input as -suffix', () => {
    const { container } = render(
      <FormControl
        formName={'formName'}
        controlProps={{
          name: 'testName',
          label: 'label',
          type: 'type',
        }}
        register={registerCallbackStub as any}
        errors={[123] as any}
      />
    )
    const label = container.querySelector('label')
    expect(label?.htmlFor).toBe('formName-testName')
    const input = container.querySelector('input')
    expect(input?.id).toBe('formName-testName')
  })
  it('prop label should be applied to label as text', () => {
    const { container } = render(
      <FormControl
        formName={'formName'}
        controlProps={{
          name: 'name',
          label: 'testLabel',
          type: 'type',
        }}
        register={registerCallbackStub as any}
        errors={[123] as any}
      />
    )
    const label = container.querySelector('label')
    expect(label?.textContent).toBe('testLabel')
  })
  it('prop type should be applied to input as type', () => {
    const { container } = render(
      <FormControl
        formName={'formName'}
        controlProps={{
          name: 'name',
          label: 'label',
          type: 'checkbox',
        }}
        register={registerCallbackStub as any}
        errors={[123] as any}
      />
    )
    const input = container.querySelector('input')
    expect(input?.type).toBe('checkbox')
  })
  it('prop placeholder should be applied to input as placeholder', () => {
    const { container } = render(
      <FormControl
        formName={'formName'}
        controlProps={{
          name: 'name',
          label: 'label',
          type: 'type',
          placeholder: 'testPlaceholder',
        }}
        register={registerCallbackStub as any}
        errors={[123] as any}
      />
    )
    const input = container.querySelector('input')
    expect(input?.placeholder).toBe('testPlaceholder')
  })
  it('input should have readonly attibute if readOnly prop true', () => {
    const { container } = render(
      <FormControl
        formName={'formName'}
        controlProps={{
          name: 'name',
          label: 'label',
          type: 'type',
        }}
        register={registerCallbackStub as any}
        errors={[123] as any}
        readOnly
      />
    )
    const input = container.querySelector('input')
    expect(input).toHaveAttribute('readonly')
  })
})
