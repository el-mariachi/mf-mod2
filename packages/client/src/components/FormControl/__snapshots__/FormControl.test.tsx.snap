// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#FormControl FormControl should to match snapshot 1`] = `
<div
  className="mb-3 row"
>
  <label
    className="form-label col-form-label col-sm-3"
    htmlFor="formName-name"
  >
    label
  </label>
  <div
    className="col-sm-9"
  >
    <input
      className="form-control"
      id="formName-name"
      type="type"
    />
    <div
      className="invalid-feedback"
    >
      undefined
    </div>
  </div>
</div>
`;
